cuda/dot_benchmark.cu

```diff
@@ -0,0 +1,197 @@
+                                     ar_data.gpu_res());
+                         },
+                         ar_get_result},
+        benchmark_info_t{"DOT Acc<fp64, fp32>",
+                         [&](matrix_info x_info, matrix_info y_info) {
+                             acc_dot(device_prop, x_info, st_data.gpu_x(),
+                                     y_info, st_data.gpu_y(),
+                                     ar_data.gpu_res());
+                         },
+                         ar_get_result},
+        benchmark_info_t{"DOT Acc<fp32, fp32>",
+                         [&](matrix_info x_info, matrix_info y_info) {
+                             acc_dot(device_prop, x_info, st_data.gpu_x(),
+                                     y_info, st_data.gpu_y(),
+                                     st_data.gpu_res());
+                         },
+                         st_get_result},
+        benchmark_info_t{"CUBLAS DOT fp64",
+                         [&](matrix_info x_info, matrix_info y_info) {
+                             cublas_dot(cublas_handle.get(), x_info,
+                                        ar_data.gpu_x(), y_info,
+                                        ar_data.gpu_y(), ar_data.gpu_res());
+                         },
+                         ar_get_result},
+        benchmark_info_t{"CUBLAS DOT fp32",
+                         [&](matrix_info x_info, matrix_info y_info) {
+                             cublas_dot(cublas_handle.get(), x_info,
+                                        st_data.gpu_x(), y_info,
+                                        st_data.gpu_y(), st_data.gpu_res());
+                         },
+                         st_get_result}};
+
+    std::cout << "Vector Size";
+    if (!detailed_error) {
+        for (const auto &info : benchmark_info) {
+            std::cout << DELIM << std::get<0>(info);
+        }
+        for (const auto &info : benchmark_info) {
+            std::cout << DELIM << "Error " << std::get<0>(info);
+        }
+    } else {
+        for (const auto &info : benchmark_info) {
+            std::cout << DELIM << "Error " << std::get<0>(info);
+        }
+    }
+    std::cout << '\n';
+
+    std::cout.precision(16);
+    std::cout << std::scientific;
+
+    auto get_error = [](value_type res, value_type ref_res) -> value_type {
+        return std::abs(res - ref_res) / std::abs(ref_res);
+    };
+    constexpr std::size_t start = std::min(max_size, std::size_t{1'000'000});
+    constexpr std::size_t row_incr = 2'000'000;
+    constexpr std::size_t steps = (max_size - start) / row_incr;
+    constexpr std::size_t randomize_num{10};
+
+    std::vector<std::size_t> benchmark_vec_size((steps + 1));
+    std::vector<double> benchmark_time((steps + 1) * benchmark_num);
+    std::vector<value_type> benchmark_error((steps + 1) * benchmark_num);
+
+    for (std::size_t randomize = 0;
+         (detailed_error && randomize < randomize_num) ||
+         (!detailed_error && randomize < 1);
```

Comment:
Yes. Additionally, when printing detailed_error, the dot function is only called once instead of the usual 10 times (since we don't print the runtime anymore).